php/class-customize-snapshot-manager.php

```diff
@@ -914,4 +973,34 @@ public function get_front_uuid_param() {
 	public function get_customize_uuid_param() {
 		return constant( get_class( $this->post_type ) . '::CUSTOMIZE_UUID_PARAM_NAME' );
 	}
+
+	/**
+	 * Publishes changeset from frontend.
+	 */
+	public function ajax_snapshot_frontend_publish() {
+		if ( ! check_ajax_referer( 'customize-snapshots-frontend-publish', 'nonce' ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'bad_nonce' );
+		}
+
+		if ( ! isset( $_POST['uuid'] ) ) {
+			return;
+		}
+
+		$this->current_snapshot_uuid = esc_attr( $_POST['uuid'] );
+		$this->ensure_customize_manager();
+		$r = $this->customize_manager->save_changeset_post( array(
+			'status' => 'publish',
+		) );
+
+		if ( is_wp_error( $r ) ) {
+			$msg = __( 'Publishing failed: ', 'customize-snapshots' );
+			foreach ( $r->errors as $name => $value ) {
+				$msg .= $name . '; ';
+			}
```

Comment:
This could be replaced with: `$msg .= join( "; ", array_keys( $r->errors ) )`. But is the it the key or the values that have the error messages to share?